// @backstage/integration/dist/gitea/GiteaIntegration.esm.js
import { basicIntegrations, defaultScmResolveUrl } from '../helpers.esm.js';
import { readGiteaConfig } from './config.esm.js';
import { getGiteaEditContentsUrl } from './core.esm.js';

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, key + "" , value);
  return value;
};
const _GiteaIntegration = class _GiteaIntegration {
  constructor(config) {
    this.config = config;
  }
  get type() {
    return "gitea";
  }
  get title() {
    return this.config.host;
  }
  resolveUrl(options) {
    return defaultScmResolveUrl(options);
  }
  resolveEditUrl(url) {
    return getGiteaEditContentsUrl(this.config, url);
  }
};
__publicField(_GiteaIntegration, "factory", ({ config }) => {
  var _a;
  const configs = (_a = config.getOptionalConfigArray("integrations.gitea")) != null ? _a : [];
  const giteaConfigs = configs.map((c) => readGiteaConfig(c));
  return basicIntegrations(
    giteaConfigs.map((c) => new _GiteaIntegration(c)),
    (gitea) => gitea.config.host
  );
});
let GiteaIntegration = _GiteaIntegration;

export { GiteaIntegration };
